MdeModulePkg/Library/BootMaintenanceManagerUiLib/UpdatePage.c

```diff
@@ -527,9 +527,12 @@ UpdateConsolePage (
         ((NewTerminalContext->IsStdErr != 0) && (UpdatePageId == FORM_CON_ERR_ID))
         )
     {
-      CheckFlags         |= EFI_IFR_CHECKBOX_DEFAULT;
-      ConsoleCheck[Index] = TRUE;
-    } else {
+      CheckFlags |= EFI_IFR_CHECKBOX_DEFAULT;
+
+      if (Index < MAX_MENU_NUMBER) {

```

Comment:
Why do we need to check this condition that was asserted on line 524?

Reply:
Asserts are a form of debug aid but cannot be used for ensuring functional correctness in production code.

The `ASSERT()` macro is defined as [this](https://github.com/tianocore/edk2/blob/b240eab03530f063ef5438497d70a731b19a201e/MdePkg/Include/Library/DebugLib.h#L395-L407)  in `DebugLib.h`:

```c
#if !defined (MDEPKG_NDEBUG)
#define ASSERT(Expression)        \
    do {                            \
      if (DebugAssertEnabled ()) {  \
        if (!(Expression)) {        \
          _ASSERT (Expression);     \
          ANALYZER_UNREACHABLE ();  \
        }                           \
      }                             \
    } while (FALSE)
#else
#define ASSERT(Expression)
#endif
```

So, it will be removed entirely if `DebugAssertEnabled()` is `FALSE`.

Every instance of `DebugLib` can implement `DebugAssertEnabled()` differently. Most (non-Null instances) follow the same pattern by depending on `DEBUG_PROPERTY_DEBUG_ASSERT_ENABLED` in `PcdDebugPropertyMask`:

```c
BOOLEAN
EFIAPI
DebugAssertEnabled (
  VOID
  )
{
  return (BOOLEAN)((PcdGet8 (PcdDebugPropertyMask) & DEBUG_PROPERTY_DEBUG_ASSERT_ENABLED) != 0);
}
```

In general, it's very common for different build types (like a `RELEASE` build) to either use [`BaseDebugLibNull`](https://github.com/tianocore/edk2/blob/b240eab03530f063ef5438497d70a731b19a201e/MdePkg/Library/BaseDebugLibNull/DebugLib.c#L155-L162), in which case it is implemented as:

```c
BOOLEAN
EFIAPI
DebugAssertEnabled (
  VOID
  )
{
  return FALSE;
}
```

Or disable asserts in the PCD.

In any case, the code cannot depend on asserts to be active to prevent fatal errors.